prow/cmd/sinker/main.go

```diff
@@ -532,3 +560,78 @@ func podNeedsKubernetesFinalizerCleanup(log *logrus.Entry, pj *prowapi.ProwJob,
 
 	return false
 }
+
+func getDuplicates(pjs *prowapi.ProwJobList) ([]string, error) {
+	type duplicationKey struct {
+		job string
+		// pointer types are only equal
+		// when they point to the same
+		// address, so we have to serialize
+		// the value of these fields
+		refs      string
+		extraRefs string
+		state     prowapi.ProwJobState
+	}
+	duplicates := map[duplicationKey][]string{}
+
+	var key duplicationKey
+	var err error
+	for _, pj := range pjs.Items {
+		if pj.Spec.Type == prowapi.PeriodicJob || !pj.Complete() {
+			continue
+		}
+
+		key.job = pj.Spec.Job
+		key.refs, err = comparableRefsString(pj.Spec.Refs)
+		if err != nil {
+			return nil, err
+		}
+		key.extraRefs, err = comparableExtraRefsString(pj.Spec.ExtraRefs)
+		if err != nil {
+			return nil, err
+		}
+		key.state = pj.Status.State
+
+		duplicates[key] = append(duplicates[key], pj.Name)
+	}
+
+	var result []string
+	for _, duplicateSlice := range duplicates {
+		if len(duplicateSlice) == 1 {
+			continue
+		}
+		result = append(result, duplicateSlice[1:]...)
```

Comment:
I assume this is based on the assumption of pjs are sorted based on start timestamp, is this always true? Just want to make sure we are not deleting latest run by accident

Reply:
Tbh I didn't think about which to keep at all :) I've added some sorting so youngest comes first so the one we keep will be the younged